### PR TITLE
add rstudio server (new formula)

### PR DESCRIPTION
**Pull request by randy3k**
as title

### CODE REVIEW

File: Library/Formula/rstudio-server.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
+class RstudioServer < Formula
+  homepage "http://www.rstudio.com"
+  url "https://github.com/rstudio/rstudio/archive/v0.98.1103.tar.gz"
+  sha256 "084049aae03cbaaa74a848f491d57cffab5ea67372ece0e256d54e04bd5fc6da"
+
+  depends_on "ant" => :build
+  depends_on "cmake" => :build
+  depends_on "r" => :build
```
Comment (mikemcquaid) [inline Library/Formula/rstudio-server.rb]: Are you sure this isn't needed at runtime?

File: Library/Formula/rstudio-server.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+  def install
+    # patch the boost install scripts
+    inreplace "dependencies/common/install-boost", "/opt/rstudio-tools/boost",
+      "#{prefix}/boost"
+    inreplace "dependencies/common/install-boost", "sudo ", ""
+    inreplace "src/cpp/CMakeLists.txt", "/opt/rstudio-tools/boost", "#{prefix}/boost"
+
+    # specific path to R explicitly
+    inreplace "dependencies/common/install-packages", "R CMD build",
```
Comment (mikemcquaid) [inline Library/Formula/rstudio-server.rb]: Can you please submit these patches to the upstream developers of this project and add a link to the upstream patch submission and explanation of why the patch is needed in a comment in the formula file. Thanks!

File: Library/Formula/rstudio-server.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    system "./install-dictionaries"
+    system "./install-mathjax"
+    system "./install-boost"
+    system "./install-packages"
+    system "./install-pandoc"
+    Dir.chdir "../.."
+
+    # the real body
+    Dir.mkdir "build"
+    Dir.chdir "build"
```
Comment (mikemcquaid) [inline Library/Formula/rstudio-server.rb]: Use `chdir do` and `mkdir do` blocks.

File: Library/Formula/rstudio-server.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    system "cmake", "..",
+      "-DRSTUDIO_TARGET=Server",
+      "-DCMAKE_BUILD_TYPE=Release",
+      "-DBOOST_ROOT=#{prefix}/boost/boost_1_50_0",
```
Comment (mikemcquaid) [inline Library/Formula/rstudio-server.rb]: Should this boost version be hardcoded?

File: Library/Formula/rstudio-server.rb [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+      "-DBoost_INCLUDE_DIR=#{prefix}/boost/boost_1_50_0/include",
+      "-DCMAKE_INSTALL_PREFIX=#{prefix}"
+    system "make", "install"
+
+    def caveats
+      <<-EOS.undent
+        Launch RStudio server with the command `rserver`,
+        then visit `http://localhost:8787` in the browser .
+      EOS
```
Comment (mikemcquaid) [inline Library/Formula/rstudio-server.rb]: Remove these caveats; they aren't Homebrew specific.

File: Library/Formula/rstudio-server.rb [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    end
+  end
```
Comment (mikemcquaid) [inline Library/Formula/rstudio-server.rb]: Can a test be added to do something more substantial than e.g. `--version` or `--help`? See `cmake.rb` for an example of an application formula with a good test and `tinyxml2.rb` for an example of a library formula with a good test. Thanks!

File: Library/Formula/rstudio-server.rb (hunk 1 above)
Comment (randy3k) [inline Library/Formula/rstudio-server.rb]: `r` is needed, but users can use different versions of R at runtime. For instance, they can install the mac binary or use macport to install R.

File: Library/Formula/rstudio-server.rb (hunk 2 above)
Comment (randy3k) [inline Library/Formula/rstudio-server.rb]: could patch do `"#{prefix}/boost"`? we need these as some paths are hardwired in the scripts.

File: Library/Formula/rstudio-server.rb (hunk 3 above)
Comment (randy3k) [inline Library/Formula/rstudio-server.rb]: will do so

File: Library/Formula/rstudio-server.rb (hunk 4 above)
Comment (randy3k) [inline Library/Formula/rstudio-server.rb]: I try linking the homebrew boost and homebrew/version boost155, but neither of them work.
the boost install script comes with `rstudio` and the version 1.50.0 is used specifically. I believe it is possible to make another PR to homebrew/version and use that formula. For now, I would like to make it easier.

File: Library/Formula/rstudio-server.rb (hunk 5 above)
Comment (randy3k) [inline Library/Formula/rstudio-server.rb]: will do so

File: Library/Formula/rstudio-server.rb (hunk 1 above)
Comment (mikemcquaid) [inline Library/Formula/rstudio-server.rb]: In that case let's make this `:recommended`.

Comment (mikemcquaid) [inline Library/Formula/rstudio-server.rb]: And/or add a R `Requirement`.

File: Library/Formula/rstudio-server.rb (hunk 2 above)
Comment (mikemcquaid) [inline Library/Formula/rstudio-server.rb]: The patch could do whatever but you'll need to work with upstream to find out what they find acceptable.

File: Library/Formula/rstudio-server.rb (hunk 4 above)
Comment (mikemcquaid) [inline Library/Formula/rstudio-server.rb]: Ok so they have a vendored boost that you're using? If so: that's cool (and good, actually).

File: Library/Formula/rstudio-server.rb (hunk 1 above)
Comment (randy3k) [inline Library/Formula/rstudio-server.rb]: However R is needed is the compile time to compile a R package, should I keep it as :recommended?

File: Library/Formula/rstudio-server.rb (hunk 2 above)
Comment (randy3k) [inline Library/Formula/rstudio-server.rb]: In fact, those are not really patches, but customization of the scripts to facilitate the installation.
Some of the paths are hardwired, eg `/opt/rstudio-tools/boost`, and I have to move it into homebrew cellar.

Comment (mikemcquaid) [inline Library/Formula/rstudio-server.rb]: Sure but it would be good to at least suggest to upstream that some of these could be CMake variables instead :smile:

File: Library/Formula/rstudio-server.rb (hunk 6 above)
Comment (randy3k) [inline Library/Formula/rstudio-server.rb]: I cannot think of a substantial test for rstudio-server, in fact, it even doesn't have a `--version` option in the command.

Comment (mikemcquaid) [inline Library/Formula/rstudio-server.rb]: We need to work out something here.

**Comment by randy3k** [comment]
I have updated the PR, please check it.

**Comment by randy3k** [comment]
The test is failed, but there are no tests!?

**Comment by tdsmith** [comment]
Yes; there's no formula for `r` in core, so it failed here: http://bot.brew.sh/job/Homebrew%20Pull%20Requests/22734/version=yosemite/console

**Comment by tdsmith** [comment]
Elaborating: `depends_on "r"` would have to be `depends_on "homebrew/science/r"`.

Do we want this in core or should we put this in homebrew/science?

**Comment by randy3k** [comment]
@tdsmith thanks for pointing it out.

I am happy to move it to homebrew/science. Any other suggestions?

**Comment by randy3k** [comment]
closed, i will move it to homebrew/science
